webhook/server.js

```diff
@@ -0,0 +1,377 @@
+                parent.usersCache.set(u.id, u)
+            })
+        }
+    }
+
+    // sanitizeUserId transform a chatKit user id to a suitable stream user id
+    sanitizeUserId(id) {
+        // todo improve me
+        return id.replace(':', '_')
+    }
+
+    // converts a chatKit user to a stream user
+    async createStreamUser(chatKitUser) {
+        await this.streamClient().updateUser(this.toStreamUser(chatKitUser))
+    }
+
+    // converts a chatKit user user to a a stream user
+    toStreamUser(chatKitUser) {
+        return {
+            id: this.sanitizeUserId(chatKitUser.id),
+            image: chatKitUser.avatar_url,// todo omit keys with empty values
+            name: chatKitUser.name,
+            ...chatKitUser.custom_data,
+        };
+    }
+
+    // converts a chatKit room to a stream channel
+    async getOrCreateRoom(room) {
+        let type = 'livestream'
+        if (room.private) {
+            type = 'messaging'
+        }
+
+        await this.ensureMembersExists(room.member_user_ids)
+        const streamChannel = this.streamClient().channel(type, room.id, {
+            name: room.name,
+            members: room.member_user_ids,
+            created_by_id: room.created_by_id,
+            ...room.custom_data
+        })
+        await streamChannel.create()
+        return streamChannel
+    }
+
+    // converts a chatKit room to a stream channel
+    async toStreamMessage(streamChannel, chatKitMessage) {
+        const streamMessage = {
+            id: chatKitMessage.id.toString(),
+            user_id: chatKitMessage.user_id,
+            text: chatKitMessage.text,
+            attachments: [],
+        }
+
+        //build attachments
+        for (let i = 0; i < chatKitMessage.parts.length; i++) {
+            const part = chatKitMessage.parts[i]
+            if (part.url) {
+                this.addUrlPart(streamMessage, part)
+            } else if (part.attachment) {
+                    await this.addAttachmentPart(streamChannel, streamMessage, part)
+            } else {
+                //inline part
+                switch (part.type) {
+                    case 'text/plain':
+                        streamMessage.text = part.content;
+                }
+            }
+        }
+
+        return streamMessage
+    }
+
+    // url parts are added to the message body and scrapped server side
+    addUrlPart(streamMessage, part) {
+        if (streamMessage.text) {
+            streamMessage.text += "\n"
+            streamMessage.text += part.url
+        } else {
+            streamMessage.text = part.url
+        }
+    }
+
+    async addAttachmentPart(streamChannel, streamMessage, part) {
+        const isImage = part.type.startsWith("image/")
+        const streamAttachment = {
+            id: part.attachment.id,
+            file_size: part.attachment.size,
+            text: part.attachment.name,
+            mime_type: part.type,
+            type: part.type.split('/')[0],
```

Comment:
any reason you're only taking first part of mime type? what happens for video/mp4 vs video/wav or image/png vs image/jpeg?

Reply:
our clients (react/ios/android) recognize video/image types and not the mime type